src/riscv/lib/src/machine_state/page_cache.rs

```diff
@@ -0,0 +1,404 @@
+//! - first, by bypassing the fetch/parse/run cycle for each instruction
+//! - secondly, by allowing hotspots of frequently re-executed sequences of instructions to be
+//!   JIT-compiled - thus running even faster.
+//!
+//! The page cache does this by populating the cache on a per-page basis. This allows the
+//! page-cache to _very strictly_ only be populated for pages that have R+X permissions set, but
+//! *are not* writable. The pleasant outcome of this is that we no longer have to keep the page
+//! cache as part of the PVM state: it is automatically always synced with data memory -- and
+//! therefore execution using the page cache is semantically identical to the fetch/parse/run
+//! cycle.
+
+use std::marker::PhantomData;
+
+use super::MachineCoreState;
+use super::ProgramCounterUpdate;
+use super::StepManyResult;
+use super::instruction::Instruction;
+use super::memory;
+use super::memory::Address;
+use super::memory::MemoryConfig;
+use crate::state_backend::ManagerBase;
+use crate::state_backend::ManagerRead;
+use crate::state_backend::ManagerReadWrite;
+use crate::traps::Exception;
+
+/// Per page, we store exactly the number of instruction halfwords we could fetch from that page's
+/// memory.
+///
+/// This is exactly half the page size - we know the instruction program counter is always
+/// halfword-aligned.
+const INSTRUCTION_ENTRIES: usize = 1
+    << memory::OFFSET_BITS
+        .checked_sub(1)
+        .expect("OFFSET_BITS is non-zero") as usize;
+
+/// Mask for indexing into a page's memory.
+const PAGE_OFFSET_MASK: u64 = (1 << memory::OFFSET_BITS) - 1;
+
+/// Instance of the page cache.
+#[expect(
+    unused,
+    reason = "PageCache is not integrated into the rest of the codebase"
+)]
+pub trait PageCache<MC: MemoryConfig, M: ManagerBase> {
+    /// Instantiate a new page cache instance.
+    fn new() -> Self;
+
+    /// Retrieve page that is dispatchable against the [`MachineCoreState`].
+    fn get_page_dispatch(&mut self, addr: Address) -> Option<PageDispatch<'_, MC, M>>
+    where
+        M: ManagerRead;
+
+    /// Populate a page with instruction and dispatch information, if the page has R+X permissions.
+    fn populate_page(&mut self, address: Address, core: &MachineCoreState<MC, M>)
+    where
+        M: ManagerReadWrite;
+
+    /// Invalidate a range of pages, usually due to the corresponding memory becoming write-able,
+    /// or no longer executable.
+    ///
+    /// While `pages` are expected to be `PAGE_SIZE` aligned, this is not a hard requirement.
+    /// Rather, all pages partially or fully overlapping with the range will be invalidated.
+    fn invalidate_range(&mut self, pages: std::ops::Range<u64>);
+}
+
+/// A page containing entrypoints that may then be dispatched against the [`MachineCoreState`].
+pub(crate) struct PageDispatch<'a, MC: MemoryConfig, M: ManagerBase> {
+    page: &'a [Instruction; INSTRUCTION_ENTRIES],
+    _pd: PhantomData<(MC, M)>,
```

Comment:
The `Instruction` itself isn't specific to `MemoryConfig` or `Manager`. How come you need capture those type variables in this type?